validation/validate.ts

```diff
@@ -1,62 +1,298 @@
     client,
     page
   );
-  if (targetNodeRef) {
-    const accnames = await runComparison(targetNodeRef, page, client);
-    console.log('AccName comparison:\n', accnames);
 
-    const htmlUsedByChrome = await getHTMLUsedByChrome(
-      targetNodeRef,
+  const comparisonResults = await runComparison(targetNodeRef, page, client);
+  await browser.close();
+  if (comparisonResults.disagrees) {
+    const caseId = saveTestcase(comparisonResults, 'output/snippet');
+    return {caseId: caseId, accnames: comparisonResults.accnames};
+  }
+
+  return {accnames: comparisonResults.accnames};
+}
+
+/**
+ * Compares AccName implementations on every Node in the DOM of the web page
+ * at the URL provided.
+ * @param url - The URL for the web page whose Nodes will be used to compare
+ * AccName implementations.
+ * @return The name of the directory containing the results of the URL comparison.
+ */
+export async function runURLComparison(url: string): Promise<string> {
+  const browser = await puppeteer.launch({
+    args: ['--enable-blink-features=AccessibilityObjectModel'],
+  });
+  const page = await browser.newPage();
+  const client = await page.target().createCDPSession();
+
+  const httpResponse = await page.goto(url);
+  if (!httpResponse?.ok()) {
+    throw new Error(`URL '${url}' could not be accessed.'`);
+  }
+
+  await loadAccNameLibraries(page);
+
+  // Trim URL string to get a concise directory name for this comparison.
+  const dirName = url.replace(/^(?:https?:\/\/)?(?:www\.)?/i, '').split('/')[0];
+  child.exec(`mkdir output/${dirName}`);
+
+  // Associates each category encountered with the number
+  // of occurrences of that category
+  const categoryCount: {[key: string]: number} = {};
+
+  const allNodes = await page.$$('body *');
+  let i = 0;
+  for (const node of allNodes) {
+    console.log(++i + '/' + allNodes.length);
+
+    await page.evaluate(
+      node => node.setAttribute('accnameComparisonTarget', 'true'),
+      node
+    );
+
+    const targetNodeRef = await getNodeRefFromSelector(
+      '[accnameComparisonTarget]',
       client,
       page
     );
-    console.log('\nHTML used by Chrome:\n', htmlUsedByChrome);
+
+    const comparisonResults = await runComparison(targetNodeRef, page, client);
+    if (comparisonResults.disagrees && comparisonResults.category) {
+      // Count category occurrences, save test case for any new categories.
+      const categoryHash = JSON.stringify(comparisonResults.category);
+      if (categoryCount[categoryHash]) {
+        categoryCount[categoryHash] += 1;
+      } else {
+        await saveTestcase(comparisonResults, `output/${dirName}`);
+        categoryCount[categoryHash] = 1;
+      }
+    }
+
+    await page.evaluate(
+      node => node.removeAttribute('accnameComparisonTarget'),
+      node
+    );
   }
 
+  // All categories encountered duirng comparison and their associated counts.
+  const categoryStats = Object.entries(categoryCount).map(entry => {
+    return {
```

Comment:
`return` is unnecessary for simple arrow functions like this - `entry => {category....` works